python/tests/test_stream.py

```diff
@@ -80,3 +80,23 @@ def test_response_exception():
     resp.flush()
     with pytest.raises(TChannelException):
         yield resp.write_body("aaaa")
+
+
+@pytest.mark.gen_test
+def test_error_during_stream(io_loop):
+    stream = InMemStream()
+
+    @pytest.mark.gen_test
+    def set_exception():
+        try:
+            1 / 0
+        except Exception as e:
+            stream.set_exception(e)
```

Comment:
Why does this need to be done in a callback at all. It's an asynchronous exception. You can set it right away and check that it's thrown when you `yield` the `read()`.